### PR TITLE
feat: introduce set type constructors

### PR DESCRIPTION
**Pull request by mlutze**
related to #3764

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala [hunk 1]
```diff
@@ -299,6 +299,41 @@ object TypeConstructor {
...
+    * A type constructor that represents the intersection of two effect sets.
+    */
+  case object Intersection extends TypeConstructor {
+    def kind: Kind = Kind.Effect ->: Kind.Effect ->: Kind.Effect
+  }
+
+  /**
+    * A type constructor that represents the difference of two effect sets.
+    */
+  case object Difference extends TypeConstructor {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala:326]: I am unsure whether we should be keeping Difference here. I guess the answer will reveal itself when we get to the Boolean Unification part.

File: main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala [hunk 2]
```diff
@@ -299,6 +299,41 @@ object TypeConstructor {
...
+    def kind: Kind = Kind.Effect ->: Kind.Effect ->: Kind.Effect
+  }
+
+  /**
+    * A type constructor that represents a single effect.
+    */
+  case class Effect(sym: Symbol.EffectSym) extends TypeConstructor {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala:333]: I think this should go before all the operations and I also think it should be a (non-empty) set.

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala [hunk 3]
```diff
@@ -2377,12 +2436,32 @@ object Resolver {
   /**
     * Returns the type `And(tpe1, tpe2)`.
     */
-  private def mkAnd(tpe1: Type, tpe2: Type, loc: SourceLocation): Type = Type.Apply(Type.Apply(Type.Cst(TypeConstructor.And, loc), tpe1, loc), tpe2, loc)
+  private def mkAnd(tpe1: Type, tpe2: Type, loc: SourceLocation): Type = Type.mkApply(Type.Cst(TypeConstructor.And, loc), List(tpe1, tpe2), loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:2439]: Any reason these do not call `Type.mkAnd` etc?

**Comment by magnus-madsen** [comment]
I added some comments, but I will merge this as-is, since these are easy to fix in a subsequent PR :)

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala (hunk 3 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:2439]: The helpers in `Type.scala` are too smart. They simplify the type and we lose information that is important for kind inference and error messages.

File: main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala (hunk 1 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/TypeConstructor.scala:326]: I think it is important to keep around for error messages. It's effectively a type alias, and we keep those through Typing

File: main/src/ca/uwaterloo/flix/language/phase/Resolver.scala (hunk 3 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Resolver.scala:2439]: OK, oversmart, if you will.

**Pull request merged**